### PR TITLE
Allowing config values to start with a newline

### PR DESCRIPTION
**Pull request by Siecje**
Fixes https://github.com/takluyver/pynsist/issues/112

If you only want to support starting with a newline for `packages`, `pypi_wheels` and `exclude` you just need to add `.strip()` before `.splitlines()` here.

https://github.com/takluyver/pynsist/blob/9cd4cd73e04d52b6b22353dce7d2e3bed9fd8f7f/nsist/__init__.py#L543-L544

### CODE REVIEW

File: nsist/tests/test_configuration_validator.py [hunk 1]
```diff
@@ -1,17 +1,75 @@
...
 
 DATA_FILES = os.path.join(os.path.dirname(__file__), 'data_files')
 
 def test_valid_config():
     configfile = os.path.join(DATA_FILES, 'valid_config.cfg')
-    configreader.read_and_validate(configfile)
+    config = configreader.read_and_validate(configfile)
+    print(config['Application'])
+    print('Application' in config)
+    print(config.has_section('Application'))
```
Comment (takluyver) [inline nsist/tests/test_configuration_validator.py]: Leftover debugging? Want to drop an `assert` in here instead?

File: nsist/tests/test_configuration_validator.py [hunk 2]
```diff
@@ -1,17 +1,75 @@
...
+def test_valid_config_with_values_starting_on_new_line():
+    configfile = os.path.join(DATA_FILES, 'valid_config_value_newline.cfg')
+    config = configreader.read_and_validate(configfile)
+    sections = ('Application', 'Python', 'Include', 'Build')
+    assert len(config.sections()) == len(sections)
+    for section in sections:
+        assert section in config
+        assert config.has_section(section)
+
+    assert config.get('Application', 'name') == '\nMy App'
```
Comment (takluyver) [inline nsist/tests/test_configuration_validator.py]: This is very thorough, but checking for all these incorrect values before transforming them seems superfluous - do you want to check for just one or two and then call `get_installer_builder_args`?

File: nsist/__init__.py [hunk 3]
```diff
@@ -500,23 +500,52 @@ def run(self, makensis=True):
...
+    args['py_version'] = config.get('Python', 'version', fallback=DEFAULT_PY_VERSION).strip()
+    args['py_bitness'] = config.getint('Python', 'bitness', fallback=DEFAULT_BITNESS)
+    args['py_format'] = config.get('Python', 'format').strip() if 'format' in config['Python'] else None
+    inc_msvcrt = config.get('Python', 'include_msvcrt', fallback='True')
+    args['inc_msvcrt'] = get_boolean(inc_msvcrt.strip())
+    args['build_dir'] = config.get('Build', 'directory', fallback=DEFAULT_BUILD_DIR).strip()
+    args['installer_name'] = config.get('Build', 'installer_name') if 'installer_name' in config['Build'] else None
+    args['nsi_template'] = config.get('Build', 'nsi_template').strip() if 'nsi_template' in config['Build'] else None
+    args['exclude'] = config.get('Include', 'exclude', fallback='').strip().splitlines()
+    return args
```
Comment (takluyver) [inline nsist/__init__.py]: Let's move this function into the configreader module, as it's really to do with the format of the config file.

**Comment by takluyver** [comment]
Thanks, this looks good, and I'm happy to allow a preceding newline on all values, though I wouldn't encourage people to use it for the single values. I've made a few specific comments inline.

File: nsist/tests/test_configuration_validator.py (hunk 2 above)
Comment (Siecje) [inline nsist/tests/test_configuration_validator.py]: It ensures that it is stripping the leading newline.

Usually long tests don't matter you write them once and leave them.

**Comment by takluyver** [comment]
Thanks!

**Comment by takluyver** [comment]
Just noticed I'm getting errors now, as shown below. I'm going to scale it back to just stripping multi-line values.

```
Traceback (most recent call last):
  File "/home/takluyver/.local/bin/pynsist", line 4, in <module>
    main()
  File "/home/takluyver/.local/lib/python3.5/site-packages/nsist/__init__.py", line 535, in main
    args = get_installer_builder_args(cfg)
  File "/home/takluyver/.local/lib/python3.5/site-packages/nsist/configreader.py", line 234, in get_installer_builder_args
    args['installer_name'] = config.get('Build', 'installer_name')
  File "/usr/lib64/python3.5/configparser.py", line 778, in get
    d = self._unify_values(section, vars)
  File "/usr/lib64/python3.5/configparser.py", line 1138, in _unify_values
    raise NoSectionError(section)
configparser.NoSectionError: No section: 'Build'
```

**Comment by Siecje** [comment]
Why doesn't your traceback have the fallback?

https://github.com/takluyver/pynsist/blob/106d50a63b23b34e4e7452e47e799d7de3c7d29b/nsist/configreader.py#L229

**Comment by takluyver** [comment]
It was from a line like [this one](https://github.com/takluyver/pynsist/pull/113/files#diff-71909709b14304e74f0713cad4e43bafR233), which does provide a fallback, but it fails if there isn't a `Build` section in the config file at all. I converted it back to the simpler way we got the fallback before this PR.

**Comment by Siecje** [comment]
Weird that the line in the traceback above is

```
args['installer_name'] = config.get('Build', 'installer_name')
```

without a safe fall back.

This line
`args['installer_name'] = config.get('Build', 'installer_name') if 'installer_name' in config['Build'] else None`

Should be

`args['installer_name'] = config.get('Build', 'installer_name', fallback=None) if 'Build' in config else None`

**Comment by takluyver** [comment]
Oh, I think that traceback was after I'd edited the line to fix the first error I saw.

The `fallback=` value works if the 'Build' section is missing as well, so it doesn't need a separate if/else.

**Pull request merged**